### PR TITLE
e2e test layout changes for vsphere (#398)

### PR DESCRIPTION
**Pull request by rjog**
**What this PR does / why we need it**:
For vSphere Cloud Provider there are many tests added. But all those tests are in e2e/storage directory.
This change is an effort to make vsphere tests modular.
New package 'vsphere' under test/e2e/storage/ is being created.

**Which issue(s) this PR fixes**:
Fixes # https://github.com/vmware/kubernetes/issues/353

**Special notes for your reviewer**:
This is a change purely for vSphere tests
Engineers in VMware have reviewed it.

**Release note**:
```release-note
None
```

### CODE REVIEW

File: hack/.golint_failures [hunk 1]
```diff
@@ -796,6 +796,7 @@ test/e2e/scalability
 test/e2e/scheduling
 test/e2e/servicecatalog
 test/e2e/storage
+test/e2e/storage/vsphere
```
Comment (saad-ali) [inline hack/.golint_failures]: Please fix the lint errors. Only generated files should be added to exception list at this point.

Comment (saad-ali) [reply hack/.golint_failures]: From @rohitjogvmw offline:

```
if I try removing dot imports I need to have something like ..
gomega.Expect(err).NotTo(gomega.HaveOccurred())
which is not readable.
I guess that is why test/e2e/storage is added in go lint exception.
We will loose the readability of the test code if we want to fix golint failures
```

This is a valid argument. I withdraw my comment.

Comment (rohitjogvmw) [reply hack/.golint_failures]: Hi Saad, currently test code uses dot imports.

. "github.com/onsi/ginkgo"
. "github.com/onsi/gomega"

golint complains about these dot imports.
If I remove the dot import then we will loose the readability of the test code.

Sample code before:
Expect(err).NotTo(HaveOccurred())

Sample code after removing dot from import:
gomega.Expect(err).NotTo(gomega.HaveOccurred())

Is it possible not to loose test code readability and also fix golint?

**Comment by saad-ali** [comment]
/assign @ixdy

To approve `./hack` change

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+const (
+	NodeStateTimeout            = 1 * time.Minute
+	KStart           KubeletOpt = "start"
+	KStop            KubeletOpt = "stop"
+	KRestart         KubeletOpt = "restart"
+)
+
+// podExec wraps RunKubectl to execute a bash cmd in target pod
+func podExec(pod *v1.Pod, bashExec string) (string, error) {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: why not use the existing `podExec` in e2e/storage/ or move that podExec to framework/util.go or another common location? It is a very small wrapper but is redundant.

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
+/*
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: why is "k8s_" in this file name?

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	return framework.RunKubectl("exec", fmt.Sprintf("--namespace=%s", pod.Namespace), pod.Name, "--", "/bin/sh", "-c", bashExec)
+}
+
+// kubeletCommand performs `start`, `restart`, or `stop` on the kubelet running on the node of the target pod and waits
+// for the desired statues..
+// - First issues the command via `systemctl`
+// - If `systemctl` returns stderr "command not found, issues the command via `service`
+// - If `service` also returns stderr "command not found", the test is aborted.
+// Allowed kubeletOps are `KStart`, `KStop`, and `KRestart`
+func kubeletCommand(kOp KubeletOpt, c clientset.Interface, pod *v1.Pod) {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: This func appears to be copied from _e2e/storage/persistent-volumes_disturptive.go_. Please call the original func or if that's not possible move the original to a more accessible location. Almost always there are  "out-of-sync" problems when code is replicated.

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	if kOp == KStart || kOp == KRestart {
+		// For kubelet start and restart operations, Wait until Node becomes Ready
+		if ok := framework.WaitForNodeToBeReady(c, pod.Spec.NodeName, NodeStateTimeout); !ok {
+			framework.Failf("Node %s failed to enter Ready state", pod.Spec.NodeName)
+		}
+	}
+}
+
+// getKubeletMainPid return the Main PID of the Kubelet Process
+func getKubeletMainPid(nodeIP string, sudoPresent bool, systemctlPresent bool) string {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: ditto re. replicated code...

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+	sshResult, err := framework.SSH(command, nodeIP, framework.TestContext.Provider)
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("SSH to Node %q errored.", nodeIP))
+	framework.LogSSHResult(sshResult)
+	Expect(sshResult.Code).To(BeZero(), "Failed to get kubelet PID")
+	Expect(sshResult.Stdout).NotTo(BeEmpty(), "Kubelet Main PID should not be Empty")
+	return sshResult.Stdout
+}
+
+// testKubeletRestartsAndRestoresMount tests that a volume mounted to a pod remains mounted after a kubelet restarts
+func testKubeletRestartsAndRestoresMount(c clientset.Interface, f *framework.Framework, clientPod *v1.Pod, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: ditto

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+	By("Testing that written file is accessible.")
+	out, err = podExec(clientPod, fmt.Sprintf("cat %s", file))
+	framework.Logf(out)
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("Volume mount detected on pod %s and written file %s is readable post-restart.", clientPod.Name, file)
+}
+
+// testVolumeUnmountsFromDeletedPod tests that a volume unmounts if the client pod was deleted while the kubelet was down.
+func testVolumeUnmountsFromDeletedPod(c clientset.Interface, f *framework.Framework, clientPod *v1.Pod, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: ditto

File: test/e2e/storage/vsphere/k8s_utils.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+	By("Expecting the volume mount not to be found.")
+	result, err = framework.SSH(fmt.Sprintf("mount | grep %s", clientPod.UID), nodeIP, framework.TestContext.Provider)
+	framework.LogSSHResult(result)
+	Expect(err).NotTo(HaveOccurred(), "Encountered SSH error.")
+	Expect(result.Stdout).To(BeEmpty(), "Expected grep stdout to be empty (i.e. no mount found).")
+	framework.Logf("Volume unmounted on node %s", clientPod.Spec.NodeName)
+}
+
+// runInPodWithVolume runs a command in a pod with given claim mounted to /mnt directory.
+func runInPodWithVolume(c clientset.Interface, ns, claimName, command string) {
```
Comment (jeffvance) [inline test/e2e/storage/vsphere/k8s_utils.go]: ditto

**Comment by jeffvance** [comment]
@saad-ali what is your opinion on my comments about duplicated code?
/lgtm cancel

**Comment by saad-ali** [comment]
> @saad-ali what is your opinion on my comments about duplicated code?

Completely agreed!! Please avoid duplicating code. Will leave this PR to @jeffvance to approve and lgtm

File: test/e2e/storage/vsphere/k8s_utils.go (hunk 2 above)
Comment (rohitjogvmw) [reply test/e2e/storage/vsphere/k8s_utils.go]: I will work on it.
I will get podExec in framework/util.go and will use it in vsphere tests and e2e/storage too.

File: test/e2e/storage/vsphere/k8s_utils.go (hunk 4 above)
Comment (rohitjogvmw) [reply test/e2e/storage/vsphere/k8s_utils.go]: If vsphere test needs to call function in e2e/storage then it is creating cyclic imports.
Is it fine if I move this function into e2e/storage/utils/utils.go and use it?

File: test/e2e/storage/vsphere/k8s_utils.go (hunk 5 above)
Comment (rohitjogvmw) [reply test/e2e/storage/vsphere/k8s_utils.go]: Same as above.
Is it ok If I move this code into e2e/storage/utils and use it?

File: test/e2e/storage/vsphere/k8s_utils.go (hunk 6 above)
Comment (rohitjogvmw) [reply test/e2e/storage/vsphere/k8s_utils.go]: ditto.

File: test/e2e/storage/vsphere/k8s_utils.go (hunk 5 above)
Comment (jeffvance) [reply test/e2e/storage/vsphere/k8s_utils.go]: Yes. The cyclic imports is a problem but moving common code to a package that is generally imported seems reasonable. So _framework/_ or _storage/_ seem like good locations.

**Comment by jeffvance** [comment]
The changes looks good. Thanks for your efforts in restructuring common code and solving the cyclic imports!
/lgtm

**Pull request merged**